.travis.yml

```diff
@@ -5,8 +5,23 @@
 
 dist: bionic
 os: linux
-language: go
-go: 1.14.4
+
+# set cache directories manually, because
+# we are using a non-standard directory struct
+# cargo root is in srs/agent
+#
+# If needed, caches can be cleared
+# by ways documented in
+# https://docs.travis-ci.com/user/caching#clearing-caches
+language: rust
+rust:
+  - 1.44.1
+  - stable
+cache:
+  cargo: true
+  directories:
+    - src/agent/target
```

Comment:
Hi @liubin - can you explain how this works for PRs which change the agent code. Presumably, the cache is somehow ignored, or recreated in that scenario?

Reply:
The cached content includes:
- $TRAVIS_HOME/.cache/sccache
- $TRAVIS_HOME/.cargo/
- $TRAVIS_HOME/.rustup/
- src/agent/target

So if there are no dependencies changed, the cache will hit and reduce the time compiling the external crates.

Rust agent and  rustjails/protocols/... will  always be forced to  build and  not use caches.